### ISSUE TITLE
Crash in optimize.c / convert_code_r()

### ISSUE THREAD

**Issue by gvanem** [the-tcpdump-group/libpcap#484]
When building libpcap with `-DBDEBUG` and running `filtertest` with a rather complex filter, I get a crash in `convert_code_r()` at the line:

```
#ifdef BDEBUG
bids[dst - conv_state->fstart] = p->id + 1;
#endif
```

The content of `test.filter`:

```
icmp || (
  udp && !(
    port 4500 ||
    (host 10.0.0.1 && port (514 || 1900 || 161)) ||
    (host 10.0.0.10 && port (514 || 1900)) ||
    (host 10.0.0.4 && port (161 || 5353)) ||
    (host 10.0.0.26 && port 80) ||
    (host 10.0.0.255 && port (137 || 138))
  )
) || (tcp && !(
    (host 10.0.0.1 && port 23) ||
    (host 217.13.6.35 && port 110) ||
    (host 217.13.6.35 && port 995) ||
    (host 217.13.9.15 && port 119)
  )
)
|| !(arp || ip)
```

Running this as `filtertest.exe -d -O -F test.filter en10mb`.
Running with no optimize (`filtertest.exe -d -F test.filter en10mb`), showed no overrun.

Tracing the code around this, I saw `convert_code_r()` was accessing `bids[]` with indices up to approx. 9000 for this filter! Increasing number of elements to 10000 fixed the crash, but for how long?

Or am I doing something silly by defining `BDEBUG` and running with such a large filter?
# 

**Edit:**
It seems `-O` turns _off_ optimization (?). If so, this is backwards IMHO.

**Comment by Lekensteyn** [comment]
I cannot reproduce this issue with libpcap-1.8.1-975-gc5ec9686, built on Arch Linux with:
```
./configure --disable-dbus --enable-yydebug --enable-optimizer-dbg \
    CFLAGS=-fsanitize=address\ -g\ -pthread LDFLAGS=-fsanitize=address\ -pthread
```

**Comment by jmayer** [comment]
I could not reproduce the problem with current git head either, but I have submitted a pull request (#691) that asserts in case the bids array would overflow.

**Comment by guyharris** [comment]
Should be fixed by 9bac75ac0222df31e29ecacf2da423702b3e5c3d, which adds bounds checks.
